From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Wed, 10 Jul 2024 15:06:30 +0800
Subject: [PATCH] Folia: Sync vehicle position to player position on player
 data load

You can find the original code on https://github.com/PaperMC/Folia

This allows the player to be re-positioned before logging into
the world without causing thread checks to trip on Folia.

diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index ab96dcb873758f9d92eecb62b473b84dc9e295a2..80ff1a11b37979b09d3972686272cd256f5a1981 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -452,10 +452,16 @@ public abstract class PlayerList {
     }
     private void mountSavedVehicle(ServerPlayer player, ServerLevel worldserver1, Optional<CompoundTag> optional) {
         // Paper end - Fire PlayerJoinEvent when Player is actually ready
-        if (optional.isPresent() && ((CompoundTag) optional.get()).contains("RootVehicle", 10)) {
-            CompoundTag nbttagcompound = ((CompoundTag) optional.get()).getCompound("RootVehicle");
+        if (optional.isPresent() && optional.get().contains("RootVehicle", 10)) {
+            CompoundTag nbttagcompound = optional.get().getCompound("RootVehicle");
             ServerLevel finalWorldServer = worldserver1; // CraftBukkit - decompile error
+            Vec3 playerPos = player.position(); // Paper - force sync root vehicle to player position
             Entity entity = EntityType.loadEntityRecursive(nbttagcompound.getCompound("Entity"), worldserver1, (entity1) -> {
+                // Paper start - force sync root vehicle to player position
+                if (entity1.distanceToSqr(player) > (5.0 * 5.0)) {
+                    entity1.setPosRaw(playerPos.x, playerPos.y, playerPos.z, true);
+                }
+                // Paper end - force sync root vehicle to player position
                 return !finalWorldServer.addWithUUID(entity1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.MOUNT) ? null : entity1; // CraftBukkit - decompile error // Paper - Entity#getEntitySpawnReason
             });
 
